Hotsapi.Uploader.Common/Manager.cs

```diff
@@ -115,39 +118,76 @@ public async void Start()
 
             _analyzer.MinimumBuild = await _uploader.GetMinimumBuild();
 
-            for (int i = 0; i < MaxThreads; i++) {
-                Task.Run(UploadLoop).Forget();
-            }
+            _ = Task.Run(() => ParseLoop());
+            _ = Task.Run(() => FingerprintLoop());
+            _ = Task.Run(() => UploadLoop());
         }
 
-        private async Task UploadLoop()
-        {
-            while (true) {
-                try {
-                    var file = await processingQueue.TakeAsync();
-
-                    file.UploadStatus = UploadStatus.InProgress;
-
-                    // test if replay is eligible for upload (not AI, PTR, Custom, etc)
-                    var replay = _analyzer.Analyze(file);
-                    if (file.UploadStatus == UploadStatus.InProgress) {
-                        // if it is, upload it
-                        await _uploader.Upload(file);
-                    }
-                    SaveReplayList();
-                    if (ShouldDelete(file, replay)) {
-                        DeleteReplay(file);
-                    }
-                }
-                catch (Exception ex) {
-                    _log.Error(ex, "Error in upload loop");
-                }
+        private async Task ParseLoop() {
+            using (var rateLimitParsing = new SemaphoreSlim(MaxThreads)) {

```

Comment:
Does this need to be wrapped in a 'while (true)'?  I think this 'ParseLoop' thread may be ending the first time there are no more replays to parse, and not starting again when new games are played while the uploader is open

Reply:
Not according to the contract of that enumerator, it should block on next until new stuff is thrown on the queue. It's a different datastructure than the one I implemented in this PR, and has no bulk dequeue (which I need for bulk fingerprint checking)